hibernate-core/src/main/java/org/hibernate/event/service/internal/EventListenerGroupImpl.java

```diff
@@ -125,118 +120,141 @@ public void addDuplicationStrategy(DuplicationStrategy strategy) {
 		duplicationStrategies.add( strategy );
 	}
 
-	/**
-	 * Implementation note: should be final for performance reasons.
-	 * @deprecated this is not the most efficient way for iterating the event listeners.
-	 * See {@link #fireEventOnEachListener(Object, BiConsumer)} and co. for better alternatives.
-	 */
 	@Override
-	@Deprecated
-	public final Iterable<T> listeners() {
-		final List<T> ls = listenersAsList;
-		return ls == null ? Collections.EMPTY_LIST : ls;
+	public void appendListener(T listener) {
+		handleListenerAddition( listener, this::internalAppend );
 	}
 
 	@Override
 	@SafeVarargs
 	public final void appendListeners(T... listeners) {
-		internalAppendListeners( listeners );
-		checkForArrayRefresh();
+		//noinspection ForLoopReplaceableByForEach
+		for ( int i = 0; i < listeners.length; i++ ) {
+			handleListenerAddition( listeners[i], this::internalAppend );
+		}
 	}
 
-	private void checkForArrayRefresh() {
-		final List<T> list = listenersAsList;
-		if ( this.listeners == null ) {
-			T[] a = (T[]) Array.newInstance( eventType.baseListenerInterface(), list.size() );
-			listeners = list.<T>toArray( a );
+	private void internalAppend(T listener) {
+		prepareListener( listener );
+
+		if ( listeners == null ) {
+			//noinspection unchecked
+			this.listeners = (T[]) Array.newInstance( eventType.baseListenerInterface(), 1 );
+			this.listeners[0] = listener;
 		}
-	}
+		else {
+			final int size = this.listeners.length;
+
+			//noinspection unchecked
+			final T[] newCopy = (T[]) Array.newInstance( eventType.baseListenerInterface(), size+1 );
```

Comment:
could we add space chars around `+` like `size + 1`?

Reply:
I prefer this style for simple arith